Core develop

Jonitan opened the pull request.

Jonitan committed:
Added use of dataclass.
Added non default init, repr & str methods.

Jonitan committed:
Moved init check to post_init.

Jonitan committed:
Added new exception types.
Moved related exceptions to different file of its own.

Jonitan committed:
Changed to new exceptions.
Fix minor bugs.
Added new class methods.
Added new operators.

Jonitan committed:
Added new exception types.

Jonitan committed:
Inserted new exception types in required methods.
Fixed minor bugs.
Added matrix multiplication by another matrix.
Updated tests (commented out).

Jonitan committed:
Added eq, ne, hash, iter & next methods.
Changed matrix multiplication operation.
Changed addition and subtraction to be more generic.
Updated TODO's.
Updated tests (commented out).

Jonitan committed:
Add init file and setup file.
First version - 1.0.0
Fix imports - 1.0.1

Jonitan committed:
Improved matrix multiplication and subtraction implementation.
Removed generic implementation - now assuming matrix of order NxN.

yairigal commented on matrix/__init__.py:
```diff
@@ -0,0 +1,3 @@
+from .matrix import Matrix
+from .matrix_exceptions import RowTypeNotTuple, RowItemNotNumber, RowAndColumnNumberDifferent, ObjectTypeNotMatrix, \
+                              ObjectTypeNotLegalToMultiply, MatrixNotLegalToMultiply, IllegalInput
```
```python 
from .matrix_exceptions import *
```

yairigal commented on matrix/matrix.py:
```diff
@@ -0,0 +1,193 @@
+from dataclasses import dataclass
+
+from matrix_exceptions import RowTypeNotTuple, RowItemNotNumber, RowAndColumnNumberDifferent, ObjectTypeNotMatrix, \
+                              ObjectTypeNotLegalToMultiply, MatrixNotLegalToMultiply, IllegalInput
```
Add all in brackets and remove \ at the end

yairigal commented on matrix/matrix.py:
```diff
@@ -0,0 +1,193 @@
+
+    matrix: tuple = ()
+
+    def __post_init__(self):
```
Whats the purpose of the` __post_init__` ?
Why not just use `__init__`?

yairigal commented on matrix/matrix.py:
```diff
@@ -0,0 +1,193 @@
+        print("%s" % (self.matrix,))
+
+    def is_valid_matrix(self):
+        for row in self.matrix:
+            if type(row) is not tuple:
+                raise RowTypeNotTuple(row)
+            for item in row:
+                if type(item) is not int and type(item) is not float:
+                    raise RowItemNotNumber(item)
+            if len(row) != len(self.matrix):
+                raise RowAndColumnNumberDifferent(row, len(self.matrix))
```
I think its way cleaner like:
```python
if any(type(row) is not tuple for row in self.matrix):
    raise TypeError("All rows should be tuples")

if any(any(type(item) is not int for item in row) for row in self.matrix):
    raise TypeError("Cells should be numbers")

if any(len(self.matrix) != len(row) for row in self.matrix):
    raise ValueError("Number of rows and columns should be the same")
```

Also, There are alot of builtin exception in python that pretty much cover all the basic uses of exceptions there is not need to create more.

yairigal commented on matrix/matrix.py:
```diff
@@ -0,0 +1,193 @@
+    matrix: tuple = ()
+
+    def __post_init__(self):
+        if self.is_valid_matrix():
+            object.__setattr__(self, "matrix", self.matrix)
+            object.__setattr__(self, "order", len(self.matrix))
+            object.__setattr__(self, "row_pointer", 0)
```
What the point of this class being dataclass?
You are abusing the code.

yairigal commented on matrix/matrix.py:
```diff
@@ -0,0 +1,193 @@
+            object.__setattr__(self, "row_pointer", 0)
+
+    def __str__(self):
+        return "%s" % (self.matrix,)
```
fstring

yairigal commented on matrix/matrix.py:
```diff
@@ -0,0 +1,193 @@
+        return "%s" % (self.matrix,)
+
+    def __repr__(self):
+        return "Matrix(%s)" % (self.matrix,)
```
What happens if we inherit from this class and implement a new one? will we want to get Matrix printed?

yairigal commented on matrix/matrix.py:
```diff
@@ -0,0 +1,193 @@
+        return "Matrix(%s)" % (self.matrix,)
+
+    def __neg__(self):
+        return Matrix(tuple(tuple(-item for item in self.matrix[j]) for j in range(self.order)))
```
iterate the rows instead of the indices

yairigal commented on matrix/matrix.py:
```diff
@@ -0,0 +1,193 @@
+
+    def __add__(self, other):
+        if type(other) is not Matrix:
+            raise ObjectTypeNotMatrix(other)
```
TypeError

yairigal commented on matrix/matrix.py:
```diff
@@ -0,0 +1,193 @@
+            raise ObjectTypeNotMatrix(other)
+
+        if self.order >= other.order:
+            return Matrix(tuple(tuple(self.matrix[j][i] + other.matrix[j][i]
+                                      if i < other.order and j < other.order else self.matrix[j][i]
+                                      for i in range(self.order)) for j in range(self.order)))
```
wtf is this, why do we ever need it in matrix

yairigal commented on matrix/matrix.py:
```diff
@@ -0,0 +1,193 @@
+                                      if i < other.order and j < other.order else self.matrix[j][i]
+                                      for i in range(self.order)) for j in range(self.order)))
+        else:
+            return Matrix(tuple(tuple(self.matrix[j][i] + other.matrix[j][i]
+                                      if i < self.order and j < self.order else other.matrix[j][i]
+                                      for i in (range(other.order))) for j in (range(other.order))))
```
Iterate over items not indices

yairigal commented on matrix/matrix.py:
```diff
@@ -0,0 +1,193 @@
+    def __mul__(self, other):
+        if type(other) is int or type(other) is float:
+            return Matrix(tuple(tuple(self.matrix[j][i] * other
+                                      for i in range(self.order)) for j in range(self.order)))
```
iterate over indices

yairigal commented on matrix/matrix.py:
```diff
@@ -0,0 +1,193 @@
+            if self.order != other.order:
+                raise MatrixNotLegalToMultiply(other)
+
+            return Matrix(tuple(tuple(sum(row[i] * cul[i] for i in range(self.order)) for cul in zip(*other))
```
iterate the items instead over i

yairigal commented on matrix/matrix.py:
```diff
@@ -0,0 +1,193 @@
+        if type(other) is not int and type(other) is not float:
+            raise ObjectTypeNotLegalToMultiply(other)
+
+        return Matrix(tuple(tuple(self.matrix[j][i] * other
+                                  for i in range(self.order)) for j in range(self.order)))
```
call the mul you already defined

yairigal commented on matrix/matrix.py:
```diff
@@ -0,0 +1,193 @@
+                                  for i in range(self.order)) for j in range(self.order)))
+
+    def __sub__(self, other):
+        if type(other) is not Matrix:
+            raise ObjectTypeNotMatrix(other)
```
If you use this alot you can make a decorator

yairigal commented on matrix/matrix.py:
```diff
@@ -0,0 +1,193 @@
+        if type(other) is not int and type(other) is not float:
+            raise ObjectTypeNotLegalToMultiply(other)
+
+        return Matrix(tuple(tuple(self.matrix[j][i] / other
+                                  for i in range(self.order)) for j in range(self.order)))
```
iterate over items

yairigal commented on matrix/matrix.py:
```diff
@@ -0,0 +1,193 @@
+                                  for i in range(self.order)) for j in range(self.order)))
+
+    def __eq__(self, other):
+        if type(other) is not Matrix:
```
well now you have to make a decorator

yairigal commented on matrix/matrix.py:
```diff
@@ -0,0 +1,193 @@
+        if self.order != other.order:
+            return False
+
+        for i in range(self.order):
+            if self.matrix[i] != other.matrix[i]:
+                return False
+
```
you can just check if the tuples are the same, the eq of tuples do what you did now

yairigal commented on matrix/matrix.py:
```diff
@@ -0,0 +1,193 @@
+    def __ne__(self, other):
+        return not self == other
+
+    def __iter__(self):
+        object.__setattr__(self, "row_pointer", 0)
+        return self
+
+    def __next__(self):
+        if self.row_pointer >= self.order:
+            raise StopIteration
+        else:
+            result = self.matrix[self.row_pointer]
+            object.__setattr__(self, "row_pointer", self.row_pointer + 1)
+            return result
```
You dont need to implement all the logic behind this, since tuples already did it for you.
tuples has their own iterator

yairigal commented on matrix/matrix.py:
```diff
@@ -0,0 +1,193 @@
+            return result
+
+    def __hash__(self):
+        return hash((tuple(self.matrix[j][i] for i in range(self.order) for j in range(self.order)), self.order))
```
dude just hash the tuples

yairigal commented on matrix/matrix.py:
```diff
@@ -0,0 +1,193 @@
+        return hash((tuple(self.matrix[j][i] for i in range(self.order) for j in range(self.order)), self.order))
+
+    @property
+    def tuples(self):
+        print("%s" % (self.matrix,))
```
You want to return the tuples, since this is a property the users want to access the value

yairigal commented on matrix/matrix.py:
```diff
@@ -0,0 +1,193 @@
+        if type(number) is not int or number < 0:
+            raise IllegalInput(number)
+
+        return cls(tuple(tuple(1 if i == j else 0 for i in range(number)) for j in range(number)))
```
nice

yairigal commented on matrix/matrix.py:
```diff
@@ -0,0 +1,193 @@
+        if type(number) is not int or number < 0:
+            raise IllegalInput(number)
+
+        return cls(tuple(tuple(1 for i in range(number)) for i in range(number)))
```
dude you know what to do here

yairigal commented on matrix/matrix.py:
```diff
@@ -0,0 +1,193 @@
+print(a + c)
+print(b + a)
+print(c + a)
+print(a + a)
+print(a + Matrix.unity(2))
+# print(a + ((1,),))
+
+print(a - b)
+print(b - a)
+print(a - c)
+print(c - a)
+print(a - a)
+print(a - Matrix.unity(2))
+# print(a - ((1,),))
+
+print(10 * a)
+print(10.1 * a)
+print(a / 10)
+print(a / 10.1)
+
+b = Matrix(((2, 2), (1, 1)))
+
+print(a * b)
+print(b * a)
+
+print(a == b)
+print(a != b)
+
+b = Matrix(((1, 1), (2, 3)))
+
+print(a == b)
+print(a != b)
+
+for i in a:
+    print(str(i))
+
+dictionary = {}
+dictionary[Matrix(((1, 1), (2, 2)))] = 1
+dictionary[Matrix(((1, 1), (2, 2)))] = 2
+dictionary[Matrix(((1, 1), (2, 3)))] = 3
```
All of this, but in other file, and in the unittest format

yairigal commented on matrix/matrix_exceptions.py:
```diff
@@ -0,0 +1,70 @@
+class ObjectTypeNotMatrix(Exception):
```
i really dont think you need this module

yairigal commented on setup.py:
```diff
@@ -0,0 +1,21 @@
+import setuptools
```
This is nice if you want to export this as a package.

Jonitan committed:
Change exception classes to TypeError and ValueError.
Refactor.
Change to regular class from dataclass.
Add decorators.

yairigal commented on matrix/matrix.py:
```diff
@@ -0,0 +1,197 @@
+#   Add documentation.
+
+
+class Matrix(tuple):
```
You dont need to inherit from tuple. 
You just an instance of tuple in the class

yairigal commented on matrix/matrix.py:
```diff
@@ -0,0 +1,197 @@
+class Matrix(tuple):
+
+    def __init__(self, matrix: tuple):
+        if Matrix.is_valid_matrix(matrix):
```
Just run this function, you dont need the return value

yairigal commented on matrix/matrix.py:
```diff
@@ -0,0 +1,197 @@
+            self.matrix = matrix
+
+    def is_matrix(func):
+        def wrapper(*xs):
+            if type(xs[1]) is not Matrix:
+                raise TypeError("Argument should be of Matrix type.")
```
```suggestion
        def wrapper(self, other, *args, **kwargs):
            if type(other) is not Matrix:
                raise TypeError("Argument should be of Matrix type.")
```

yairigal commented on matrix/matrix.py:
```diff
@@ -0,0 +1,197 @@
+        if Matrix.is_valid_matrix(matrix):
+            self.matrix = matrix
+
+    def is_matrix(func):
```
```suggestion
    def matrix_argument_valid(func):
```

yairigal commented on matrix/matrix.py:
```diff
@@ -0,0 +1,197 @@
+        return len(self.matrix)
+
+    def __str__(self):
+        return "%s" % (self.matrix,)
```
```suggestion
        return str(self.matrix)
```

yairigal commented on matrix/matrix.py:
```diff
@@ -0,0 +1,197 @@
+        return "%s" % (self.matrix,)
+
+    def __repr__(self):
+        return "Matrix(%s)" % (self.matrix,)
```
```suggestion
        return f"Matrix({self.matrix})"
```

yairigal commented on matrix/matrix.py:
```diff
@@ -0,0 +1,197 @@
+    def __init__(self, matrix: tuple):
+        if Matrix.is_valid_matrix(matrix):
+            self.matrix = matrix
+
```
You can make a generic decorator
like that:
```python
@validate_other_type([int, float])
def __truediv__(self, other):
   ...

@validate_other_type([Matrix, int, float])
def __mul__(self, other):
   ...
```

yairigal commented on matrix/matrix.py:
```diff
@@ -0,0 +1,197 @@
+
+    @property
+    def tuples(self):
+        print("%s" % (self.matrix,))
```
```suggestion
        return self.matrix
```

Jonitan committed:
Refactor.
Update version.

Jonitan committed:
Add test_matrix file in new tests directory.
Update requirements with pytest.

Jonitan committed:
Added basic constructor argument tests.

Jonitan committed:
Updated tests.
Refactor.

Jonitan committed:
Add unity and ones test.
Add fixture for comparing between matrixs using rows.
Fix type check in unity and ones class methods.

Jonitan committed:
Added unit tests for eq, ne, add, sub, neg methods.
Updated test for test_unity.

Jonitan committed:
Added tests for mul, rmul, truediv methods.
Fixed according to test.

idoazzz commented on matrix/matrix.py:
```diff
@@ -0,0 +1,176 @@
+class Matrix:
+
+    def __init__(self, matrix: tuple):
+        Matrix.is_valid_matrix(matrix)
```
Use "self" instead of Matrix (here and everywhere) it will be more generic. (Tomorrow we will change the name of Matrix to YoniBayoniMatrix, it will be boomer to change all to occurrences of Matrix explicitly)

idoazzz commented on matrix/matrix.py:
```diff
@@ -0,0 +1,176 @@
+        return self.matrix[i]
+
+    def __hash__(self):
+        return hash(row for row in self)
```
Hash self.matrix instead, will be easier

idoazzz commented on matrix/matrix.py:
```diff
@@ -0,0 +1,176 @@
+        return hash(row for row in self)
+
+    def __neg__(self):
+        return Matrix(tuple(tuple(-item for item in row) for row in self))
```
You can use __mul__

idoazzz commented on matrix/matrix.py:
```diff
@@ -0,0 +1,176 @@
+        return not self == other
+
+    def __add__(self, other):
+        if type(other) is not Matrix:
+            raise TypeError("Argument is not of Matrix type.")
```
Those lines repeating in many places. Consider use function

yairigal replied:
I suggested a decorator

idoazzz commented on matrix/matrix.py:
```diff
@@ -0,0 +1,176 @@
+        if type(other) is not Matrix:
+            raise TypeError("Argument is not of Matrix type.")
+
+        if len(self) != len(other):
+            raise ValueError("Matrix must be of the same order.")
```
Also

idoazzz commented on matrix/matrix.py:
```diff
@@ -0,0 +1,176 @@
+        if len(self) != len(other):
+            raise ValueError("Matrix must be of the same order.")
+
+        if self is other:
+            return self * 2
```
It's redundant, there are many special cases, I don't encouraging check each case, you will calculate all those cases in the next lines.
Also, it's not more fast or optimized (the mul is more complicated calculation).

idoazzz commented on matrix/matrix.py:
```diff
@@ -0,0 +1,176 @@
+            return self * 2
+
+        return Matrix(tuple(tuple((sum(cell_pair) for cell_pair in zip(s_row, o_row)))
+                            for s_row, o_row in zip(self, other)))
```
Why making your variables names more shorter here (there are more several places)? What it gives you? 
Consider:
`return self(tuple(tuple((sum(cell_pair) for cell_pair in zip(self_row, other_row)))
                            for self_row, other_row in zip(self, other)))`

idoazzz commented on matrix/matrix.py:
```diff
@@ -0,0 +1,176 @@
+        return self * other
+
+    def __truediv__(self, other: Union[int, float]):
+        return Matrix(tuple(tuple(cell / other for cell in row) for row in self))
```
You can use __mul__

idoazzz commented on setup.py:
```diff
@@ -0,0 +1,21 @@
+        "Operating System :: OS Independent",
+    ],
+    python_requires='>=3.7',
+)
```
Nice code! good job you ass hole

Jonitan committed:
Change truediv and neg to sue mul.
Change init to use self instead of Matrix with is_valid method.
Change hash.
Update tests.

Jonitan committed:
Fix tests to be with constant pre-evaluated expected results.
Fix few bugs accordingly in matrix.
Minor refactor.

Jonitan committed:
Merge pull request #3 from Jonitan/add_tests

Add tests

Jonitan committed:
Version 1.1.0

yairigal commented on matrix/matrix.py:
```diff
@@ -0,0 +1,121 @@
+from typing import Union
```
module docs

yairigal commented on matrix/matrix.py:
```diff
@@ -0,0 +1,121 @@
+from typing import Union
+
+
+class Matrix:
+
```
class and attributes docs

yairigal commented on matrix/matrix.py:
```diff
@@ -0,0 +1,121 @@
+        return str(self.matrix)
+
+    def __repr__(self):
+        return f"Matrix({self.matrix})"
```
All the classes who will inherit from this class will also print "Matrix"

yairigal commented on matrix/matrix.py:
```diff
@@ -0,0 +1,121 @@
+
+        return self + (-other)
+
+    def __mul__(self, other):
+        if type(other) is int or type(other) is float:
```
I still suggest using a decorator to specify the types that are eligible for this function.
then you can remove all the type checking inside the function and keep only the logic.

yairigal commented on setup.py:
```diff
@@ -0,0 +1,21 @@
+import setuptools
+
+with open("README.md", "r") as fh:
+    long_description = fh.read()
```
Not sure this will work in real world. You need to add the README in the package as well

yairigal commented on tests/test_matrix.py:
```diff
@@ -0,0 +1,271 @@
+@pytest.fixture
+def compare_matrix_rows():
+    def _compare_matrix_rows(matrix_a, matrix_b):
+        for row_a, row_b in zip(matrix_a, matrix_b):
+            assert row_a == row_b
```
```suggestion
       assert matrix_a.tuples == matrix_b.tuples
```

yairigal commented on tests/test_matrix.py:
```diff
@@ -0,0 +1,271 @@
+                         list(zip(ILLEGAL_TYPE_MATRIX_LIST, [TypeError] * len(ILLEGAL_TYPE_MATRIX_LIST))) +
+                         list(zip(ILLEGAL_TYPE_MATRIX_CELL, [TypeError] * len(ILLEGAL_TYPE_MATRIX_CELL))) +
+                         list(zip(ILLEGAL_VALUE_MATRIX_LIST, [ValueError] * len(ILLEGAL_VALUE_MATRIX_LIST))))
+def test_matrix(matrix_factory, expected_matrix, expected_exception):
```
```suggestion
def test_matrix_initialization(matrix_factory, expected_matrix, expected_exception):
```

yairigal commented on tests/test_matrix.py:
```diff
@@ -0,0 +1,271 @@
+
+
+########################################################################################################################
+@pytest.mark.parametrize('expected_matrix, expected_exception',
+                         list(zip(LEGAL_MATRIX_LIST, [None] * len(LEGAL_MATRIX_LIST))) +
+                         list(zip(ILLEGAL_TYPE_MATRIX_LIST, [TypeError] * len(ILLEGAL_TYPE_MATRIX_LIST))) +
+                         list(zip(ILLEGAL_TYPE_MATRIX_CELL, [TypeError] * len(ILLEGAL_TYPE_MATRIX_CELL))) +
+                         list(zip(ILLEGAL_VALUE_MATRIX_LIST, [ValueError] * len(ILLEGAL_VALUE_MATRIX_LIST))))
```
This is way too complicated.
test should be stupid.
make the test a simple as possible, if you really need add another tests for each case. You dont need to worry about code duplication

yairigal commented on tests/test_matrix.py:
```diff
@@ -0,0 +1,271 @@
+
+
+########################################################################################################################
+@pytest.mark.parametrize('expected_matrix', LEGAL_MATRIX_LIST)
```
The whole constants for each case is confusing, i want to see what are the values you are testing.

Jonitan committed:
Minor fix for repr.
Fix explicity of uint tests.
